### PR TITLE
Patelyash/index refactor

### PR DESCRIPTION
**Pull request by yashpatel007**
- [ ] Does this PR have a descriptive title that could go in our release notes?
- [ ] Does this PR add any new dependencies?
- [ ] Does this PR modify any existing APIs?
   - [ ] Is the change to the API backwards compatible?
- [ ] Should this result in any changes to our documentation, either updating existing docs or adding new ones?

### CODE REVIEW

File: include/index_factory.h [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    bool load_on_build = false;
+
+    std::string data_type;
+    std::string label_type;
+    std::string tag_type;
+};
+
+struct BuildParams
+{
+    uint32_t R;
```
Comment (daxpryce) [inline include/index_factory.h:49]: opening discussion here: what say everyone to using more descriptive names for these parameters?  I took a stab at naming them in #300 for the old way of providing params, but I'm open to name changes from the ones I picked. 

I like single char variables in whitepapers, but I have a strong dislike for them when I'm trying to process the code I'm reading, especially if I'm not super familiar.

File: include/index_factory.h [hunk 2]
```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (daxpryce) [inline include/index_factory.h:1]: Is the idea to have an IndexFactory and a DynamicIndexFactory, or should we figure out how the dynamic indices come into here?  Is it two enums, or a singular set of enums like `DYNAMIC_DISK` and `STATIC_MEMORY`.  Also, how does Disk and SSD differ?

File: include/index_factory.h (hunk 1 above)
Comment (yashpatel007) [reply include/index_factory.h:49]: I agree, and this was created with just POC. There is so much that we can improve here. First off reading those vars from a yml config file would be great.  Then as your #300 PR can be very helpful here.  I am planning to add yml-cpp library to do the config reading.

File: include/index_factory.h (hunk 2 above)
Comment (yashpatel007) [reply include/index_factory.h:1]: I was thinking more like, we have a flag if the build is dynamic or not, while creating the Instance of Disk or Memory Index we can use this flag to appropriately instantiate the class.  Let me look into this, thanks for pointing this out.

**Comment by harsha-simhadri** [comment]
replaced with #311

**Pull request closed without merging**